# MULTI LAYER PERCEPTRON CHO BÀI TOÁN REGRESSION

## Bai tap 1

In [1]:
import numpy as np
import pandas as pd
import time

from sklearn.model_selection import train_test_split
from sklearn.metrics import r2_score, mean_squared_error
from sklearn.linear_model import LinearRegression

In [2]:
# cost or loss function
def cost(Y, Yhat):
    return np.mean((Yhat - Y) ** 2)

In [3]:
path = 'D:\Machine Learning\ANN_MLP\data\SAT_GPA.csv'

df = pd.read_csv(path)
df.head()

<>:1: SyntaxWarning: invalid escape sequence '\M'
<>:1: SyntaxWarning: invalid escape sequence '\M'
C:\Users\Admin\AppData\Local\Temp\ipykernel_13592\163357733.py:1: SyntaxWarning: invalid escape sequence '\M'
  path = 'D:\Machine Learning\ANN_MLP\data\SAT_GPA.csv'


SAT   GPA
0  1714  2.40
1  1664  2.52
2  1760  2.54
3  1685  2.74
4  1693  2.83

In [4]:
X = df["SAT"]
y = df["GPA"]

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=30/84, random_state=1)

In [5]:
X_train = X_train.to_numpy().reshape(-1, 1)
X_test = X_test.to_numpy().reshape(-1, 1)

y_train = y_train.to_numpy().reshape(-1, 1)
y_test = y_test.to_numpy().reshape(-1, 1)

In [6]:
d0 = 1
d1 = h = 100 # size of hidden layer
d2 = C = 1
# initialize parameters randomly
W1 = 0.01*np.random.randn(d0, d1)
b1 = np.zeros((d1, 1))
W2 = 0.01*np.random.randn(d1, d2)
b2 = np.zeros((d2, 1))

N = X_train.T.shape[1]
eta = 0.0005 # learning rate

start_time = time.time()
for i in range(10000):
    ## Feedforward
    Z1 = np.dot(W1.T, X_train.T) + b1
    A1 = np.where(Z1 > 0, Z1, 0.01 * Z1)  # LeakyReLU
    Z2 = np.dot(W2.T, A1) + b2
    Yhat = Z2

    # print loss after each 1000 iterations
    if i %1000 == 0:
        # compute the loss: average cross-entropy loss
        loss = cost(y_train, Yhat)
        print("iter %d, loss: %f" %(i, loss))

    # backpropagation
    E2 = (Yhat - y_train.T)/N
    dW2 = np.dot(A1, E2.T)
    db2 = np.sum(E2, axis = 1, keepdims = True)
    E1 = np.dot(W2, E2)
    E1 = np.where(Z1 > 0, E1, 0.01 * E1)  # Gradient of LeakyReLU
    dW1 = np.dot(X_train.T, E1.T)
    db1 = np.sum(E1, axis = 1, keepdims = True)
    
    # Gradient clipping
    clip_value = 1.0
    dW1 = np.clip(dW1, -clip_value, clip_value)
    dW2 = np.clip(dW2, -clip_value, clip_value)
    db1 = np.clip(db1, -clip_value, clip_value)
    db2 = np.clip(db2, -clip_value, clip_value)
    
    # Gradient Descent update
    W1 += -eta*dW1
    b1 += -eta*db1
    W2 += -eta*dW2
    b2 += -eta*db2

training_time = time.time() - start_time
print(f"\nTraining Time: %f s" % training_time)

iter 0, loss: 17.104267
iter 1000, loss: 0.201638
iter 2000, loss: 0.199882
iter 3000, loss: 0.199811
iter 4000, loss: 0.199809
iter 5000, loss: 0.199805
iter 6000, loss: 0.199804
iter 7000, loss: 0.199807
iter 8000, loss: 0.199801
iter 9000, loss: 0.199800

Training Time: 0.810584 s


In [7]:
Z1 = np.dot(W1.T, X_train.T) + b1
A1 = np.maximum(Z1, 0)
Z2 = np.dot(W2.T, A1) + b2

Z2

array([[3.14674276, 3.04632643, 2.81725166, 3.09182758, 3.03377438,
        3.215779  , 2.88628789, 2.71526631, 3.04789543, 2.65721312,
        2.92394401, 3.07770653, 2.87530485, 3.0981036 , 2.96160014,
        3.11065564, 2.82509668, 2.63995406, 2.70271427, 3.05103344,
        2.80783762, 2.78116453, 2.85333878, 3.09496559, 2.69957626,
        2.622695  , 2.79842359, 2.78900956, 2.97415218, 3.11850067,
        2.91609899, 2.98356622, 2.66035113, 2.66976516, 3.18439889,
        2.79528558, 2.82823469, 3.18282989, 2.79842359, 2.90354695,
        2.77645752, 2.84078673, 2.87844286, 2.78116453, 2.80313061,
        2.62426401, 2.76704349, 2.98356622, 2.63367804, 3.17498486,
        2.91609899, 3.11222465, 2.73566338, 3.07927554]])

In [8]:
num_runs = 10
prediction_times = []

for _ in range(num_runs):
    start_time = time.time()
    
    Z1 = np.dot(W1.T, X_train.T) + b1
    A1 = np.maximum(Z1, 0)
    Z2 = np.dot(W2.T, A1) + b2
    
    prediction_time = time.time() - start_time
    prediction_times.append(prediction_time)

avg_prediction_time = np.mean(prediction_times)
print("Average Predict Time: %f s" % avg_prediction_time)

Average Predict Time: 0.000000 s


In [9]:
print("R square:", r2_score(y_train, Z2[0]))
print("MSE:", mean_squared_error(y_train, Z2[0]))

R square: -1.4795317998992772
MSE: 0.19696836006220456


In [10]:
d0 = 1
d1 = h = 100 # size of hidden layer
d2 = C = 1
# initialize parameters randomly
W1 = 0.01*np.random.randn(d0, d1)
b1 = np.zeros((d1, 1))
W2 = 0.01*np.random.randn(d1, d2)
b2 = np.zeros((d2, 1))

N = X_test.T.shape[1]
eta = 0.0005 # learning rate

start_time = time.time()

for i in range(10000):
    ## Feedforward
    Z1 = np.dot(W1.T, X_test.T) + b1
    A1 = np.where(Z1 > 0, Z1, 0.01 * Z1)  # LeakyReLU
    Z2 = np.dot(W2.T, A1) + b2
    Yhat = Z2

    # print loss after each 1000 iterations
    if i %1000 == 0:
        # compute the loss: average cross-entropy loss
        loss = cost(y_test, Yhat)
        print("iter %d, loss: %f" %(i, loss))

    # backpropagation
    E2 = (Yhat - y_test.T)/N
    dW2 = np.dot(A1, E2.T)
    db2 = np.sum(E2, axis = 1, keepdims = True)
    E1 = np.dot(W2, E2)
    E1 = np.where(Z1 > 0, E1, 0.01 * E1)  # Gradient of LeakyReLU
    dW1 = np.dot(X_test.T, E1.T)
    db1 = np.sum(E1, axis = 1, keepdims = True)

    # Gradient clipping
    clip_value = 1.0
    dW1 = np.clip(dW1, -clip_value, clip_value)
    dW2 = np.clip(dW2, -clip_value, clip_value)
    db1 = np.clip(db1, -clip_value, clip_value)
    db2 = np.clip(db2, -clip_value, clip_value)
    
    # Gradient Descent update
    W1 += -eta*dW1
    b1 += -eta*db1
    W2 += -eta*dW2
    b2 += -eta*db2

training_time = time.time() - start_time
print(f"\nTraining Time: %f s" % training_time)

iter 0, loss: 21.095445
iter 1000, loss: 0.185336
iter 2000, loss: 0.201420
iter 3000, loss: 0.201344
iter 4000, loss: 0.197007
iter 5000, loss: 0.198850
iter 6000, loss: 0.199326
iter 7000, loss: 0.199369
iter 8000, loss: 0.199334
iter 9000, loss: 0.199357

Training Time: 1.005340 s


In [11]:
Z1 = np.dot(W1.T, X_test.T) + b1
A1 = np.maximum(Z1, 0)
Z2 = np.dot(W2.T, A1) + b2

Z2

array([[2.99142999, 3.33730941, 3.26257164, 3.20000049, 3.53892756,
        3.48678494, 3.24345268, 3.06442966, 3.37554733, 2.98273955,
        3.17914344, 3.21042902, 3.11831038, 3.19131005, 2.81588314,
        3.16002448, 3.33730941, 3.11831038, 3.14090552, 2.9340731 ,
        3.21738137, 3.03488217, 3.42942805, 3.18957197, 3.20347667,
        3.22954798, 3.33209515, 3.29038104, 3.08007245, 3.43464231]])

In [12]:
num_runs = 10
prediction_times = []

for _ in range(num_runs):
    start_time = time.time()
    
    Z1 = np.dot(W1.T, X_test.T) + b1
    A1 = np.maximum(Z1, 0)
    Z2 = np.dot(W2.T, A1) + b2
    
    prediction_time = time.time() - start_time
    prediction_times.append(prediction_time)

avg_prediction_time = np.mean(prediction_times)
print("Average Predict Time: %f s" % avg_prediction_time)

Average Predict Time: 0.000000 s


In [13]:
print("R square:", r2_score(y_test, Z2[0]))
print("MSE:", mean_squared_error(y_test, Z2[0]))

R square: -0.5269318344583531
MSE: 0.08242785224228712


In [14]:
linR = LinearRegression()

start_time = time.time()

linR.fit(X_train, y_train)

training_time = time.time() - start_time
print("Training Time: %f s" % training_time)


num_runs = 10
prediction_times = []

for _ in range(num_runs):
    start_time = time.time()
    y_pred = linR.predict(X_test)
    prediction_time = time.time() - start_time
    prediction_times.append(prediction_time)
    
avg_prediction_time = np.mean(prediction_times)
print("Average Predict Time: %f s" % avg_prediction_time)

print("R square:", r2_score(y_test, y_pred))
print("MSE:", mean_squared_error(y_test, y_pred))

Training Time: 0.002528 s
Average Predict Time: 0.000099 s
R square: 0.05950572997060655
MSE: 0.05077038868090673


In [16]:
d0 = 1
d1 = h = 75 # size of hidden layer
d2 = C = 1
# initialize parameters randomly
W1 = 0.01*np.random.randn(d0, d1)
b1 = np.zeros((d1, 1))
W2 = 0.01*np.random.randn(d1, d2)
b2 = np.zeros((d2, 1))

N = X_train.T.shape[1]
eta = 0.0005 # learning rate

start_time = time.time()

for i in range(10000):
    ## Feedforward
    Z1 = np.dot(W1.T, X_train.T) + b1
    A1 = np.where(Z1 > 0, Z1, 0.01 * Z1)  # LeakyReLU
    Z2 = np.dot(W2.T, A1) + b2
    Yhat = Z2

    # print loss after each 1000 iterations
    if i %1000 == 0:
        # compute the loss: average cross-entropy loss
        loss = cost(y_train, Yhat)
        print("iter %d, loss: %f" %(i, loss))

    # backpropagation
    E2 = (Yhat - y_train.T)/N
    dW2 = np.dot(A1, E2.T)
    db2 = np.sum(E2, axis = 1, keepdims = True)
    E1 = np.dot(W2, E2)
    E1 = np.where(Z1 > 0, E1, 0.01 * E1)  # Gradient of LeakyReLU
    dW1 = np.dot(X_train.T, E1.T)
    db1 = np.sum(E1, axis = 1, keepdims = True)
    
    # Gradient clipping
    clip_value = 1.0
    dW1 = np.clip(dW1, -clip_value, clip_value)
    dW2 = np.clip(dW2, -clip_value, clip_value)
    db1 = np.clip(db1, -clip_value, clip_value)
    db2 = np.clip(db2, -clip_value, clip_value)

     # Gradient Descent update
    W1 += -eta*dW1
    b1 += -eta*db1
    W2 += -eta*dW2
    b2 += -eta*db2

training_time = time.time() - start_time
print(f"\nTraining Time: %f s" % training_time)


iter 0, loss: 23.427248
iter 1000, loss: 0.220017
iter 2000, loss: 0.220260
iter 3000, loss: 0.220303
iter 4000, loss: 0.220310
iter 5000, loss: 0.220310
iter 6000, loss: 0.220309
iter 7000, loss: 0.220308
iter 8000, loss: 0.220307
iter 9000, loss: 0.220305

Training Time: 0.754600 s


In [18]:
Z1 = np.dot(W1.T, X_test.T) + b1
A1 = np.maximum(Z1, 0)
Z2 = np.dot(W2.T, A1) + b2

print("R square:", r2_score(y_test, Z2[0]))
print("MSE:", mean_squared_error(y_test, Z2[0]))

R square: -0.7796977051784229
MSE: 0.09607282798607839


In [20]:
d0 = 1
d1 = h = 50 # size of hidden layer
d2 = C = 1
# initialize parameters randomly
W1 = 0.01*np.random.randn(d0, d1)
b1 = np.zeros((d1, 1))
W2 = 0.01*np.random.randn(d1, d2)
b2 = np.zeros((d2, 1))

N = X_train.T.shape[1]
eta = 0.0005 # learning rate

start_time = time.time()

for i in range(10000):
    ## Feedforward
    Z1 = np.dot(W1.T, X_train.T) + b1
    A1 = np.where(Z1 > 0, Z1, 0.01 * Z1)  # LeakyReLU
    Z2 = np.dot(W2.T, A1) + b2
    Yhat = Z2

    # print loss after each 1000 iterations
    if i %1000 == 0:
        # compute the loss: average cross-entropy loss
        loss = cost(y_train, Yhat)
        print("iter %d, loss: %f" %(i, loss))

    # backpropagation
    E2 = (Yhat - y_train.T)/N
    dW2 = np.dot(A1, E2.T)
    db2 = np.sum(E2, axis = 1, keepdims = True)
    E1 = np.dot(W2, E2)
    E1 = np.where(Z1 > 0, E1, 0.01 * E1)  # Gradient of LeakyReLU
    dW1 = np.dot(X_train.T, E1.T)
    db1 = np.sum(E1, axis = 1, keepdims = True)
    
    # Gradient clipping
    clip_value = 1.0
    dW1 = np.clip(dW1, -clip_value, clip_value)
    dW2 = np.clip(dW2, -clip_value, clip_value)
    db1 = np.clip(db1, -clip_value, clip_value)
    db2 = np.clip(db2, -clip_value, clip_value)

     # Gradient Descent update
    W1 += -eta*dW1
    b1 += -eta*db1
    W2 += -eta*dW2
    b2 += -eta*db2

training_time = time.time() - start_time
print(f"\nTraining Time: %f s" % training_time)


iter 0, loss: 8.485891
iter 1000, loss: 0.161161
iter 2000, loss: 0.159679
iter 3000, loss: 0.156982
iter 4000, loss: 0.151789
iter 5000, loss: 0.150593
iter 6000, loss: 0.150436
iter 7000, loss: 0.150417
iter 8000, loss: 0.150415
iter 9000, loss: 0.150416

Training Time: 0.682132 s


In [21]:
Z1 = np.dot(W1.T, X_test.T) + b1
A1 = np.maximum(Z1, 0)
Z2 = np.dot(W2.T, A1) + b2

print("R square:", r2_score(y_test, Z2[0]))
print("MSE:", mean_squared_error(y_test, Z2[0]))

R square: -2.063773134785624
MSE: 0.16539064387742072


### Kết quả:
- Loss giảm từ 17.1 xuống 0.199 trên tập train
- Loss giảm từ 21.1 xuống 0.199 trên tập test
- R-square âm (-1.48) cho thấy mô hình không fit tốt với dữ liệu
- MSE khoảng 0.197 trên cả tập train và test

## Bai tap 2

In [22]:
import numpy as np
import pandas as pd

from sklearn.metrics import r2_score, mean_squared_error
from sklearn.linear_model import LinearRegression

In [23]:
# cost or loss function
def cost(Y, Yhat):
    return np.mean((Yhat - Y) ** 2)

In [25]:
path = 'D:\\Machine Learning\\ANN_MLP\\data\\vidu4_lin_reg.txt'

df = pd.read_csv(path, sep=" ")
df.head()

ID  TUOI  BMI   HA  GLUCOSE  CHOLESTEROL  BEDAYNTM
0   1    56   21  160     14.0         6.00      1.95
1   2    76   18  150     12.0         4.97      1.33
2   3    63   16  160      4.4         6.39      0.83
3   4    78   20  100      4.0         7.00      2.00
4   5    87   20  110      4.6         4.10      1.30

In [26]:
X = df.drop(['ID', 'BEDAYNTM'], axis=1)
y = df['BEDAYNTM']

X_train = X[:80]
y_train = y[:80]

X_test = X[80:]
y_test = y[80:]

In [27]:
y_train = y_train.to_numpy().reshape(-1, 1)
y_test = y_test.to_numpy().reshape(-1, 1)

In [29]:
d0 = 5
d1 = h = 100 # size of hidden layer
d2 = C = 1
# initialize parameters randomly
W1 = 0.01*np.random.randn(d0, d1)
b1 = np.zeros((d1, 1))
W2 = 0.01*np.random.randn(d1, d2)
b2 = np.zeros((d2, 1))

N = X_train.T.shape[1]
eta = 0.001 # learning rate

for i in range(10000):
    ## Feedforward
    Z1 = np.dot(W1.T, X_train.T) + b1
    A1 = np.where(Z1 > 0, Z1, 0.01 * Z1)  # LeakyReLU
    Z2 = np.dot(W2.T, A1) + b2
    Yhat = Z2

    # print loss after each 1000 iterations
    if i %1000 == 0:
        # compute the loss: average cross-entropy loss
        loss = cost(y_train, Yhat)
        print("iter %d, loss: %f" %(i, loss))

    # backpropagation
    E2 = (Yhat - y_train.T)/N
    dW2 = np.dot(A1, E2.T)
    db2 = np.sum(E2, axis = 1, keepdims = True)
    E1 = np.dot(W2, E2)
    E1 = np.where(Z1 > 0, E1, 0.01 * E1)  # Gradient of LeakyReLU
    dW1 = np.dot(X_train.T, E1.T)
    db1 = np.sum(E1, axis = 1, keepdims = True)
    
    # Gradient clipping (To avoid booming gradient)
    clip_value = 1.0
    dW1 = np.clip(dW1, -clip_value, clip_value)
    dW2 = np.clip(dW2, -clip_value, clip_value)
    db1 = np.clip(db1, -clip_value, clip_value)
    db2 = np.clip(db2, -clip_value, clip_value)

    # Gradient Descent update
    W1 += -eta*dW1
    b1 += -eta*db1
    W2 += -eta*dW2
    b2 += -eta*db2

print("Training done!")

iter 0, loss: 1.633611
iter 1000, loss: 0.151597
iter 2000, loss: 0.151730
iter 3000, loss: 0.151366
iter 4000, loss: 0.151597
iter 5000, loss: 0.151653
iter 6000, loss: 0.151699
iter 7000, loss: 0.151683
iter 8000, loss: 0.151657
iter 9000, loss: 0.151744
Training done!


In [30]:
Z1 = np.dot(W1.T, X_train.T) + b1
A1 = np.maximum(Z1, 0)
Z2 = np.dot(W2.T, A1) + b2

Z2

array([[1.12050537, 1.24396561, 1.07393314, 1.08829821, 1.19939812,
        1.17435056, 1.02479531, 1.0568369 , 1.17307447, 1.22543004,
        0.67791447, 1.14196214, 1.21657454, 1.02297691, 0.9733986 ,
        0.75010389, 1.07934872, 1.23229472, 1.0845406 , 1.06253063,
        1.25021193, 1.1678294 , 1.36092958, 0.8112222 , 1.34817212,
        1.2080271 , 0.99572062, 0.95518788, 1.25983538, 0.97087815,
        1.01859184, 1.13043695, 1.13477176, 0.99869171, 0.85983173,
        1.2021141 , 1.04086783, 1.06070685, 1.19347677, 1.11299237,
        1.34518202, 1.03054771, 0.77596548, 0.9837873 , 1.0180409 ,
        1.21911725, 0.93128345, 0.98178786, 1.01733744, 0.7579376 ,
        1.01506994, 1.12206541, 1.02654034, 1.09602115, 1.2470717 ,
        1.13005256, 0.79653497, 1.04417429, 0.76326444, 1.06952102,
        0.98530297, 1.0671505 , 0.97772298, 1.08913447, 0.93789676,
        0.88776262, 0.99944114, 1.08253036, 0.88229572, 1.0663376 ,
        1.17059732, 1.4595678 , 1.12507892, 0.93

In [31]:
print("R square:", r2_score(y_train, Z2[0]))
print("MSE:", mean_squared_error(y_train, Z2[0]))

R square: 0.1896031980110695
MSE: 0.10452790454648944


In [32]:
Z1 = np.dot(W1.T, X_test.T) + b1
A1 = np.maximum(Z1, 0)
Z2 = np.dot(W2.T, A1) + b2

Z2

array([[0.90317482, 0.80014592, 1.22404712, 0.94763144, 0.9499854 ,
        1.04730146, 1.2112276 , 1.0507365 , 0.99744585, 0.69406933,
        0.80628671, 0.75806581, 1.18092909, 1.06551095, 0.98182391,
        1.0264556 , 1.1521737 , 1.18551   , 1.30530844, 1.28139224]])

In [33]:
print("R square:", r2_score(y_test, Z2[0]))
print("MSE:", mean_squared_error(y_test, Z2[0]))

R square: 0.2110390308347614
MSE: 0.22841425982569344


In [34]:
linR = LinearRegression()

linR.fit(X_train, y_train)

y_pred = linR.predict(X_test)

print("R square:", r2_score(y_test, y_pred))
print("MSE:", mean_squared_error(y_test, y_pred))

R square: 0.21453893938640334
MSE: 0.22740099167615907


### Kết quả:
- R-square trên tập train đạt 0.19 và tập test đạt 0.21, cho thấy mô hình có khả năng dự đoán tốt hơn so với bài tập 1
- MSE trên tập train là 0.10 và tập test là 0.23, thể hiện mô hình có độ chính xác khá tốt
- Không có hiện tượng overfitting rõ ràng do hiệu suất trên tập test và train khá tương đồng

## Bai tap 3

In [35]:
import numpy as np
import pandas as pd

from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error

In [36]:
# cost or loss function
def cost(Y, Yhat):
    return np.mean((Yhat - Y) ** 2)

In [37]:
path = 'D:\\Machine Learning\\ANN_MLP\data\\real_estate.csv'

df = pd.read_csv(path)
df.head()

<>:1: SyntaxWarning: invalid escape sequence '\d'
<>:1: SyntaxWarning: invalid escape sequence '\d'
C:\Users\Admin\AppData\Local\Temp\ipykernel_13592\3567143085.py:1: SyntaxWarning: invalid escape sequence '\d'
  path = 'D:\\Machine Learning\\ANN_MLP\data\\real_estate.csv'


No  X1 transaction date  X2 house age  \
0   1             2012.917          32.0   
1   2             2012.917          19.5   
2   3             2013.583          13.3   
3   4             2013.500          13.3   
4   5             2012.833           5.0   

   X3 distance to the nearest MRT station  X4 number of convenience stores  \
0                                84.87882                               10   
1                               306.59470                                9   
2                               561.98450                                5   
3                               561.98450                                5   
4                               390.56840                                5   

   X5 latitude  X6 longitude  Y house price of unit area  
0     24.98298     121.54024                        37.9  
1     24.98034     121.53951                        42.2  
2     24.98746     121.54391                        47.3  
3     24.98746     121.54391                        54.8  
4     24.97937     121.54245                        43.1

In [38]:
X = df.drop(['No', 'Y house price of unit area'], axis=1)
y = df['Y house price of unit area']

X_train = X[:350]
y_train = y[:350]

X_test = X[350:]
y_test = y[350:]

y_train = y_train.to_numpy().reshape(-1, 1)
y_test = y_test.to_numpy().reshape(-1, 1)

In [39]:
d0 = 6
d1 = h = 100 # size of hidden layer
d2 = C = 1
# initialize parameters randomly
W1 = 0.01*np.random.randn(d0, d1)
b1 = np.zeros((d1, 1))
W2 = 0.01*np.random.randn(d1, d2)
b2 = np.zeros((d2, 1))

N = X_train.T.shape[1]
eta = 0.0001 # learning rate

for i in range(10000):
    ## Feedforward
    Z1 = np.dot(W1.T, X_train.T) + b1
    A1 = np.where(Z1 > 0, Z1, 0.01 * Z1)  # LeakyReLU
    Z2 = np.dot(W2.T, A1) + b2
    Yhat = Z2

    # print loss after each 1000 iterations
    if i %1000 == 0:
        # compute the loss: average cross-entropy loss
        loss = cost(y_train, Yhat)
        print("iter %d, loss: %f" %(i, loss))

    # backpropagation
    E2 = (Yhat - y_train.T)/N
    dW2 = np.dot(A1, E2.T)
    db2 = np.sum(E2, axis = 1, keepdims = True)
    E1 = np.dot(W2, E2)
    E1 = np.where(Z1 > 0, E1, 0.01 * E1)  # Gradient of LeakyReLU
    dW1 = np.dot(X_train.T, E1.T)
    db1 = np.sum(E1, axis = 1, keepdims = True)
    
    # Gradient clipping (To avoid booming gradient)
    clip_value = 1.0
    dW1 = np.clip(dW1, -clip_value, clip_value)
    dW2 = np.clip(dW2, -clip_value, clip_value)
    db1 = np.clip(db1, -clip_value, clip_value)
    db2 = np.clip(db2, -clip_value, clip_value)
    
    # Gradient Descent update
    W1 += -eta*dW1
    b1 += -eta*db1
    W2 += -eta*dW2
    b2 += -eta*db2

print("Training done!!!")

iter 0, loss: 1645.156576
iter 1000, loss: 293.227067
iter 2000, loss: 297.379269
iter 3000, loss: 299.453560
iter 4000, loss: 300.430561
iter 5000, loss: 300.968239
iter 6000, loss: 301.230568
iter 7000, loss: 301.251704
iter 8000, loss: 301.479591
iter 9000, loss: 301.459114
Training done!!!


In [40]:
Z1 = np.dot(W1.T, X_test.T) + b1
A1 = np.maximum(Z1, 0)
Z2 = np.dot(W2.T, A1) + b2

Z2

array([[43.01756405, 27.08717658, 23.10195793, 27.17269877, 27.83695955,
        47.59017742, 49.14957722, 49.8476156 , 52.28512235, 26.01553402,
        46.98958241, 40.6959884 , 32.14508341, 46.69102302, 35.23460809,
        24.51788373, 26.22103371, 26.05270577, 44.32851536, 24.15796331,
        46.53869927, 48.98821524, 45.00753869, 50.41043014, 47.05566103,
        27.16633922, 26.42965087, 54.62622889, 35.37681655, 50.50692873,
        46.99874612, 52.13627831, 20.90712866, 38.69317942, 19.40776595,
        50.70084672, 52.13880895, 25.11672307, 25.79466467, 44.10734486,
        40.93287296, 27.28286785, 36.82053419, 32.18397802, 16.30481492,
        40.52050686, 27.89747713, 29.40163501, 26.42965087, 49.4121477 ,
        39.85596968, 26.45864039, 49.05129094, 45.8449472 , 46.42980619,
        48.14854954, 48.49280638, 26.07939572, 24.50219292, 19.79925781,
        53.57192873, 43.9253967 , 52.28811164, 53.37930671]])

In [41]:
print("MSE:", mean_squared_error(y_test, Z2[0]))

MSE: 52.64472422573521


In [42]:
linR = LinearRegression()

linR.fit(X_train, y_train)

y_pred = linR.predict(X_test)

print("MSE:", mean_squared_error(y_test, y_pred))

MSE: 65.18991450477287


1. Hiệu suất huấn luyện:
- Mô hình MLP được huấn luyện trong 10000 vòng lặp với learning rate = 0.0001
- Loss giảm đáng kể từ 1645.15 xuống còn 301.45, cho thấy mô hình đã hội tụ tốt
- Việc sử dụng LeakyReLU và gradient clipping giúp tránh được vấn đề vanishing gradient

2. So sánh với Linear Regression:
- MLP cho kết quả MSE trên tập test là 52.64
- Linear Regression cho kết quả MSE trên tập test là 65.18
- MLP cho hiệu suất tốt hơn khoảng 19% so với Linear Regression

3. Đánh giá tổng quan:
- MLP thể hiện khả năng học tốt hơn các mối quan hệ phi tuyến trong dữ liệu bất động sản
- Tuy nhiên, giá trị MSE vẫn còn khá cao, có thể cải thiện thêm bằng cách:
  + Tăng số lượng hidden layer
  + Điều chỉnh learning rate
  + Thử nghiệm các hàm activation khác
  + Áp dụng các kỹ thuật regularization